internal/controllers/virtual_devices_controller.go

```diff
@@ -0,0 +1,210 @@
+			Name:              "DIMO",
+			Version:           "1",
+			ChainId:           math.NewHexOrDecimal256(vc.Settings.DeviceMintingChainID),
+			VerifyingContract: vc.Settings.DeviceMintingVerifyingContract,
+		},
+		Message: signer.TypedDataMessage{
+			"integrationNode": math.NewHexOrDecimal256(integrationID),
+			"vehicleNode":     math.NewHexOrDecimal256(vehicleNode),
+		},
+	}
+}
+
+// GetVirtualDeviceMintingPayload godoc
+// @Description gets the payload for to mint virtual device given an integration token ID
+// @Tags        integrations
+// @Produce     json
+// @Success     200 {array} signer.TypedData
+// @Router      /integration/:tokenID/mint-virtual-device [get]
+func (vc *VirtualDeviceController) GetVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	tokenID := c.Params("tokenID")
+	vehicleNode := c.Query("vehicle_id")
+	userID := helpers.GetUserID(c)
+
+	uTokenID, err := strconv.ParseUint(tokenID, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
+		Id: userID,
+	})
+	if err != nil {
+		vc.log.Debug().AnErr("GetUser Grpc", err).Msg("error occurred when fetching user")
+		return fiber.NewError(fiber.StatusUnauthorized, "error occurred when fetching user")
+	}
+
+	if user.EthereumAddress == nil {
+		return fiber.NewError(fiber.StatusUnauthorized, "User does not have an Ethereum address on file.")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "failed to get integration")
+	}
+
+	vid, err := strconv.ParseInt(vehicleNode, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid vehicleNode provided")
+	}
+	vnID := types.NewNullDecimal(decimal.New(vid, 0))
+	vehicleNFT, err := models.VehicleNFTS(
+		models.VehicleNFTWhere.TokenID.EQ(vnID),
+		models.VehicleNFTWhere.OwnerAddress.EQ(null.BytesFrom(common.HexToAddress(*user.EthereumAddress).Bytes())),
+	).Exists(c.Context(), vc.DBS().Reader)
+	if err != nil {
+		vc.log.Error().Err(err).Str("vehicleNode", vehicleNode).Str("tokenID", tokenID).Msg("Could not fetch minting payload for device")
+		return fiber.NewError(fiber.StatusInternalServerError, "error generating device mint payload")
+	}
+
+	if !vehicleNFT {
+		return fiber.NewError(fiber.StatusBadRequest, "user does not own vehicle node")
+	}
+
+	response := vc.getVirtualDeviceMintPayload(int64(integration.TokenId), vid)
+
+	return c.JSON(response)
+}
+
+// SignVirtualDeviceMintingPayload godoc
+// @Description validate signed signature for vehicle minting
+// @Tags        integrations
+// @Produce     json
+// @Success     200 {array}
+// @Router      /integration/:tokenID/mint-virtual-device [post]
+func (vc *VirtualDeviceController) SignVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	tokenID := c.Params("tokenID")
+	vehicleNode := c.Query("vehicle_id")
```

Comment:
why would `vehicle_id` not be in the path and instead in the querystring - is it optional?

Reply:
3. `POST virtual-device/mint/{integrationId}/{vehicleId}` (which order?) is also a possibility. The advantage here is that it's easy to search in Nginx, which is powerful.

Reply:
Good point, will make the change

Reply:
I am iffy on this as a URL pattern.